src/game/common/rts/team.cpp

```diff
@@ -0,0 +1,815 @@
+            }
+
+            iter.Advance();
+        }
+    }
+}
+
+void TeamFactory::Clear()
+{
+    std::map<NameKeyType, TeamPrototype *> temp(m_prototypes);
+    m_prototypes.clear();
+
+    for (std::map<NameKeyType, TeamPrototype *>::iterator it = temp.begin(); it != temp.end(); ++it) {
+        it->second->Delete_Instance();
+    }
+}
+
+void TeamFactory::Init_From_Sides(SidesList *sides)
+{
+    Clear();
+
+    for (int i = 0; i < sides->Get_Num_Teams(); i++) {
+        Dict *d = sides->Get_Team_Info(i)->Get_Dict();
+        Utf8String name = d->Get_AsciiString(g_teamNameKey);
+        Utf8String owner = d->Get_AsciiString(g_teamOwnerKey);
+        bool is_singleton = d->Get_Bool(g_teamIsSingletonKey);
+        Init_Team(name, owner, is_singleton, d);
+    }
+}
+
+void TeamFactory::Init_Team(const Utf8String &name, const Utf8String &owner, bool is_singleton, Dict *d)
+{
+    captainslog_dbgassert(!Find_Team_Prototype(name), "team already exists");
+    Player *player = g_thePlayerList->Find_Player_With_NameKey(Name_To_Key(owner));
+    captainslog_dbgassert(player != nullptr, "no owner found for team %s (%s)", owner.Str(), name.Str());
+
+    if (player == nullptr) {
+        player = g_thePlayerList->Get_Neutral_Player();
+    }
+
+    new TeamPrototype(this, name, player, is_singleton, d, ++m_nextPrototypeID);
+
+    if (is_singleton) {
+        Create_Inactive_Team(name);
+    }
+}
+
+void TeamFactory::Add_Team_Prototype_To_List(TeamPrototype *team)
+{
+    Utf8String name = team->Get_Name();
+    NameKeyType key = Name_To_Key(name);
+    std::map<NameKeyType, TeamPrototype *>::iterator it = m_prototypes.find(key);
+
+    if (it != m_prototypes.end()) {
+        captainslog_dbgassert(
+            it->second == team, "TeamFactory::Add_Team_Prototype_To_List: Team %s already exists... skipping.", name);
+    } else {
+        m_prototypes[key] = team;
+    }
+}
+
+void TeamFactory::Remove_Team_Prototype_From_List(TeamPrototype *team)
+{
+    Utf8String name = team->Get_Name();
+    NameKeyType key = Name_To_Key(name);
+    std::map<NameKeyType, TeamPrototype *>::iterator it = m_prototypes.find(key);
+
+    if (it != m_prototypes.end()) {
+        m_prototypes.erase(it);
+    }
+}
+
+TeamPrototype *TeamFactory::Find_Team_Prototype(const Utf8String &name)
+{
+    NameKeyType key = Name_To_Key(name);
+    std::map<NameKeyType, TeamPrototype *>::iterator it = m_prototypes.find(key);
+
+    if (it != m_prototypes.end()) {
+        return it->second;
+    } else {
+        return nullptr;
+    }
+}
+
+TeamPrototype *TeamFactory::Find_Team_Prototype_By_ID(unsigned int id)
+{
+    std::map<NameKeyType, TeamPrototype *>::iterator it;
+
+    for (it = m_prototypes.begin(); it != m_prototypes.end(); ++it) {
+        if (it->second->Get_ID() == id) {
+            return it->second;
+        }
+    }
+
+    return nullptr;
+}
+
+Team *TeamFactory::Find_Team_By_ID(unsigned int id)
+{
+    if (id == 0) {
+        return nullptr;
```

Comment:
This check is sure correct? Teams in game can be 0 visually, 0 to 9, but maybe it is different to this logic here?

Reply:
This is correct to the original.